types.cc

```diff
@@ -1512,6 +1512,18 @@ static void validate_aux(const tuple_type_impl& t, View v, cql_serialization_for
         }
         ++ti;
     }
+
+    size_t extra_elements = 0;
+    while (!v.empty()) {
+        read_tuple_element(v);
+        extra_elements += 1;
+    }
+
+    if (extra_elements > 0) {
+        // This function is called for both tuple and user_type, print the name too
+        throw marshal_exception(format("Value of type {} contained too many fields (expected {}, got {})",
+                                t.name(), t.size(), t.size() + extra_elements));
+    }
```

Comment:
Why this change?

Reply:
When evaluating a `bind_variable` we need to check that the serialized value represents the expected type.
This is done by calling `abstract_type::validate`.

For some reason `user_types::marker::bind` doesn't use `validate`.
[user_types::marker::bind](https://github.com/scylladb/scylla/blob/dd9d6c081ee48e670eda70e14a765442db62c0ec/cql3/user_types.cc#L141)  calls [user_types::value user_types::value::from_serialized](https://github.com/scylladb/scylla/blob/dd9d6c081ee48e670eda70e14a765442db62c0ec/cql3/user_types.cc#L57) which deserializes the elements and checks that there are no more than `size()` of them. 

That would mean that there is actually no validation for fields of user types value (???)

I wanted to have all validation done in `validate`, so I added checking number of fields to `abstract_type::validate` for `tuple_type_impl` (which also handles `user_type_impl` thanks to inheritance)

Reply:
Please check if there is indeed no validation.

Reply:
yup, I made an example:
```rust
use scylla::{Session, SessionBuilder};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let session: Session = SessionBuilder::new()
        .known_node("127.0.0.1:9042")
        .build()
        .await?;

    session.query("CREATE KEYSPACE IF NOT EXISTS ks WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1}", &[]).await?;
    session.query("CREATE TYPE IF NOT EXISTS ks.list_type (a frozen<list<int>>)", &[]).await?;
    session.query("CREATE TABLE IF NOT EXISTS ks.list_table (p int primary key, l list_type)", &[]).await?;

    let prepared = session.prepare("INSERT INTO ks.list_table (p, l) values (0, ?)").await?;

    // Insert tuple<int> instead of tuple<list<int>> as a value of type list_type, DB should refuse
    session.execute(&prepared, ((1234,),)).await?;

    Ok(())
}
```
This example finishes without errors and successfully inserts an `int32` in place of a `list<int>`
Then when I try to select using `cqlsh` I get:
```
cqlsh> select * from ks.list_table;
Traceback (most recent call last):
  File "/usr/local/bin/cqlsh", line 1042, in perform_simple_statement
    result = future.result()
  File "cassandra/cluster.py", line 4894, in cassandra.cluster.ResponseFuture.result
    raise self._final_exception
DriverException: Failed decoding result column "l" of type frozen<list_type>: Buffer slice out of bounds

```

At first I tried to make an example using python driver, but it turns out that python driver does type validation on its own and it refused to even send the query.

On this branch the example fails, although the error message is a bit ugly:
```
Error: DbError(Invalid, "Exception while binding value of type org.apache.cassandra.db.marshal.UserType(ks,6c6973745f74797065,61:org.apache.cassandra.db.marshal.FrozenType(org.apache.cassandra.db.marshal.ListType(org.apache.cassandra.db.marshal.Int32Type))): marshaling error: read_simple - not enough bytes (expected 4, got 0) Backtrace: 0x23c3b6e 0x23c4060 0x23c4368 0xcb4e5f 0xcb4ad4 0xcb4973 0xcb47d4 0x13d832a 0x13bd931 0x13bdb67 0x13c5906 0x11c08f9 0x11d46b1 0x11c02fb 0x11c045c 0x152152d 0x1297fc2 0x12983cd 0x1283226 0x1287df1 0x128b651 0x1282651 0x12846eb 0x1283773 0x128c257 0x128c7fc 0x1289804 0x224a115 0x228936f 0x228a577 0x22899ba 0x2241198 0x2240746 0xb94b65 /lib64/libc.so.6+0x27b74 0xb93a6d\n   --------\n   seastar::lambda_task<seastar::execution_stage::flush()::$_5>")
```